DESKTOP : Added a Markdown-it plugin to support embedded youtube videos

Rishgod opened the pull request:
A youtube link is converted into an embedded video which could be played in the preview pane

#### ON ENABLING THE PLUGIN

```md
[](https://www.youtube.com/watch?v=I8GtygCoNcY)
```

is interpreted as

```html
<div>
	<iframe 
	src="https://www.youtube.com/embed/I8GtygCoNcY"
	frameborder="0"
	allowfullscreen/>
</div>
```
**If you want to hide the link, wrap it in the link tag with no heading else the link or the link-tag heading would be displayed below the video player as shown below**
<br/>

![image](https://user-images.githubusercontent.com/54576074/79131375-25d60100-7dc6-11ea-8ae9-42e9f4f33f0c.png)

<br/>
fixes #2706 

@PackElend can you please label me, thanks!

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/embedvideo.js:
```diff
@@ -0,0 +1,109 @@
+
+    if (state.src.charCodeAt(state.pos) !== 0x5b /* [ */) {
+      return false;
+    }
```
That solution is not optimum as you're unnecessarily parsing the whole Markdown document even though it's already been parsed by markdown-it. Markdown-it will produce a stream of tokens, which you can hook into to provide alternative rendering. In your case, you could hook in link_open and, if you detect that the href is a YouTube URL, display your iframe code. Please have a look at the markdown-it doc for more info.

Rishgod replied:
I'll look into the same!

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/embedvideo.js:
```diff
@@ -0,0 +1,109 @@
+    '"  frameborder="0" allowfullscreen allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"></iframe></div>';
+
+  return embedStart + videoID + embedEnd;
+}
```
"tokenize" means converting a string to a stream of tokens. That's not what this function does.

Rishgod replied:
I'm writing a post that explains everything, I'll be done with it in an hour and I'll update the description of the PR for better understanding how everything works since this rule is heavily based from the repo I mentioned, I also have some gaps in my knowledge about this topic, but I'll try to explain it most of what I understand(after 2 days of research in the same)

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -15,6 +15,7 @@ const rules = {
 	fountain: require('./MdToHtml/rules/fountain'),
 	mermaid: require('./MdToHtml/rules/mermaid').default,
 	sanitize_html: require('./MdToHtml/rules/sanitize_html').default,
+	video: require('./MdToHtml/rules/embedvideo'),
```
rule should be called "video", not "embedvideo".

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/embedvideo.js:
```diff
@@ -0,0 +1,109 @@
+'use strict';
```
Normally this should be implemented in TypeScript, as mention in CONTRIBUTING.md. But let's say in JS is fine too, provided the implementation is correct.

laurent22 commented:
Thanks for looking into it. Honestly that plugin you looked at was not a good starting point. What you should look at is the markdown-it doc. In fact, I've just realised there's a good example there that you can copy almost verbatim to implement what you need.

Rishgod commented:
> Thanks for looking into it. Honestly that plugin you looked at was not a good starting point. What you should look at is the markdown-it doc. In fact, I've just realised there's a good example there that you can copy almost verbatim to implement what you need.
@laurent22 are you referencing to this? 

<img width="742" alt="Annotation 2020-03-13 224608" src="https://user-images.githubusercontent.com/54576074/76644181-7fa39b00-657c-11ea-9fb6-f60aff4aebce.png">

laurent22 commented:
Yes, except that in your case you need to modify the link_open rule instead of the image one.

Rishgod commented:
> Yes, except that in your case you need to modify the link_open rule instead of the image one.

Alright, I'll try to solve this using the above approach.

laurent22 commented:
Also please note that we'll need the following for this pull request:

- Unit tests (in tests/MdToHtml)
- Updated doc in readme/markdown.md

Rishgod commented:
Okay, I'll do the same!

Rishgod commented:
@laurent22 well there is one change in behaviour with only modifying link_open, right now we are parsing the type=inline and looking for a link type block `[youtube](URL...)`, and check if the title block has _**youtube**_, and then parse the following, if we only work on link_open all youtube links would be automatically converted into an embedded video, there would be no additional flag except for turning the plugin off **_(like in the previous case we can change the title from youtube to something else to stop this behaviour)_**

<img width="887" alt="Annotation 2020-03-14 032655" src="https://user-images.githubusercontent.com/54576074/76662374-c658bb80-65a3-11ea-9956-b4566e36955f.png">

Rishgod commented:
#### **Changes :**

![embedVideoII](https://user-images.githubusercontent.com/54576074/76664935-09b72800-65ac-11ea-886b-2d2228a03b1d.gif)

mic704b commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -367,6 +367,7 @@ class Setting extends BaseModel {
 			'markdown.plugin.multitable': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['mobile', 'desktop'], label: () => _('Enable multimarkdown table extension') },
 			'markdown.plugin.fountain': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['mobile', 'desktop'], label: () => _('Enable Fountain syntax support') },
 			'markdown.plugin.mermaid': { value: true, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['mobile', 'desktop'], label: () => _('Enable Mermaid diagrams support') },
+			'markdown.plugin.video': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['desktop'], label: () => _('Enable embededing video') },
```
You've got a small typo.  Perhaps:  'Enable video embedding'

Rishgod replied:
@mic704b thank you for pointing it out, I'm working on the creating a unit test, I'll change this when I'm done with that

Rishgod commented:
> Also please note that we'll need the following for this pull request:
> 
> * Unit tests (in tests/MdToHtml)
> * Updated doc in readme/markdown.md

@laurent22 @mic704b 
Can you please guide me a bit in writing the unit test case, right now I am basing it on according to this [unit test](https://github.com/eloquence/markdown-it-html5-media/blob/master/test/test.js)

<img width="709" alt="img" src="https://user-images.githubusercontent.com/54576074/76682088-f002f880-661e-11ea-93c6-10227ae8059d.png">





<img width="703" alt="Annotation 2020-03-14 175834" src="https://user-images.githubusercontent.com/54576074/76681950-c9908d80-661d-11ea-85c7-ea874112315e.png">

-  I have a test case file which exports object array into  MDVideoTest

-  MDVideoTest file is based on [this unit test](https://github.com/eloquence/markdown-it-html5-media/blob/master/test/test.js)

- there is plugin [markdown-it-testgen](https://www.npmjs.com/package/markdown-it-testgen) which could be downloaded from npm and it outputs the parsed cases in the .txt file. (Let me know if you require this or this should be left out)
_**"This package parses fixtures in commonmark spec format and generates tests for markdown-it parser and plugins."**_

Let me know if I have to do something  _**different**_ and this is not _**whats required**_.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,31 @@
+			}
+		} catch (e) { return ''; }
+
+		// try-catch to remove uncaught error promises when match[1] is undefined
```
Please don't silence exceptions, it leads to poor quality software since errors get undetected and never get fixed.

Secondly, the way I understand this code is like this: if the user adds a YouTube URL, display a video player. If they add any other URL, display nothing. Or am I missing something? You didn't add automated-tests yet but did you test with regular URLs in the app?

Rishgod replied:
yes, it checks if the domain name is youtube and then creates an iframe for the link, the `VideoIDRegex` extracts the 11 digit code which identifies a the youtube video ID and then concatenate it.

laurent22 replied:
> Secondly, the way I understand this code is like this: if the user adds a YouTube URL, display a video player. If they add any other URL, display nothing. Or am I missing something? You didn't add automated-tests yet but did you test with regular URLs in the app?

Rishgod replied:
Yes there is an inconsistency with regular links, 

With plugin disabled

<img width="1128" alt="Annotation 2020-03-22 193102" src="https://user-images.githubusercontent.com/54576074/77251330-c6c41700-6c73-11ea-8d19-8ac3dd8f3248.png">


Plugin enabled

<img width="1128" alt="link" src="https://user-images.githubusercontent.com/54576074/77251332-c9267100-6c73-11ea-8eac-a9011bd97468.png">


what do you think is causing this?

laurent22 commented:
Regarding testing, I mentioned tests/MdToHtml, did you look at this and understand what it does?

Rishgod commented:
> Regarding testing, I mentioned tests/MdToHtml, did you look at this and understand what it does?

well I did not look into it properly, I'll check again

Rishgod commented:
@mic704b can you please guide me in writing the integration test for this

mic704b commented:
> can you please guide me in writing the integration test for this

I don't know how you would write an integration test for this.  Perhaps you can start with unit tests as Laurent suggested.  Existing tests are in `CliClient/tests/MdToHtml.js`, they might give you inspiration, but I'm not too familiar with this area.

Rishgod commented:
> > can you please guide me in writing the integration test for this
> 
> I don't know how you would write an integration test for this. Perhaps you can start with unit tests as Laurent suggested. Existing tests are in `CliClient/tests/MdToHtml.js`, they might give you inspiration, but I'm not too familiar with this area.

Noted.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,29 @@
+		if (match && videoID[2].length == 11) {
+			const embedStart = '<div><iframe  width="560" height="315" src="https://www.youtube.com/embed/';
+			const embedEnd = '"  frameborder="0" allowfullscreen allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"></iframe></div>';
+			return embedStart + videoID[2] + embedEnd;
```
Please use a template literal to format and indent the HTML string - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

Any reason why you chose width 560? I guess it makes sense to set a min and max width, but otherwise could you make it shrink and grow with the container?

Could you document these properties and are they necessary? `accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture` For example, for me the video doesn't even auto-play (and I don't think it should) so the autoplay property maybe is not needed. I don't know about the others.

Rishgod replied:
> Any reason why you chose width 560? I guess it makes sense to set a min and max width, but otherwise could you make it shrink and grow with the container?


`width ="560" and height="315"` is to maintain the aspect ratio 16:9 which is the default aspect ratio for the majority of videos and is used to prevent black bars,


**Autoplay** means that once you finish the video should the next one autoplay(_**auto-playing** the video in youtube recommendation queue, not the one embedded in the player_)

**Picture-in-picture** shrinks the video into a small player that you can move around your home screen and position over other apps. You can continue watching YouTube videos while using other apps on your mobile device with this feature turned on.


**accelerometer and gyroscope** 
This usually refers to when a user rotates or moves their device around, so I'm guessing this would be for when someone embeds a Virtual Reality or 360degree video?

**Encrypted media** 
Encrypted Media Extensions provides an API that enables web applications to interact with content protection systems, to allow playback of encrypted audio and video

laurent22 replied:
> if we would use max-width, min-width we might have black bars in the video player because of improper aspect ratio,

Why? The height should also grow and shrink with the width, thus the ratio is not affected. Also "Any reason why you chose width 560?"

Regarding the extra property, I get the feeling we could remove them all and it wouldn't change much, or am I missing something? I guess you copied that from somewhere else, which is fine, but in our case if we don't need them, we can remove them.

Rishgod replied:
>Regarding the extra property, I get the feeling we could remove them all and it wouldn't change much, or am I missing something? I guess you copied that from somewhere else, which is fine, but in our case if we don't need them, we can remove them.

Yes it is written on the template youtube provides for embedding videos


<img width="1127" alt="Annotation 2020-03-24 164157" src="https://user-images.githubusercontent.com/54576074/77419507-8937c900-6dee-11ea-9907-40a06362ae27.png">


``` htm

<iframe width="560" height="315" src="https://www.youtube.com/embed/vdHJjbHwR38" 
frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in
-picture" allowfullscreen></iframe>

```

Rishgod replied:
> Why? The height should also grow and shrink with the width, thus the ratio is not affected. Also "Any reason why you chose width 560?"

Okay, dimension 560X315 =~ 16:9 (which is desired since most videos are in that aspect ratio) 

we can change it to width to max-width in iframe and calculate the height so it is appx 16:9,
I guess that would be a better implementation, what do you think?


#### [Current 560X315]
![Screenshot (10)](https://user-images.githubusercontent.com/54576074/77424082-8ccf4e00-6df6-11ea-837c-29e257de6401.png)

#### Changing width to 640~max width lead to black bars in video if we do not correct the aspect ratio in iframe
![Screenshot (11)](https://user-images.githubusercontent.com/54576074/77424094-8fca3e80-6df6-11ea-9336-0e54009dcb5f.png)


####  Correcting the aspect ratio
![Screenshot (12)](https://user-images.githubusercontent.com/54576074/77424126-9953a680-6df6-11ea-8d69-ba2487f6437f.png)

sinkuu replied:
> it is written on the template youtube provides for embedding videos, so no additional reason for using width 560 besides the fact it was provided by youtube when I checked the specification for the embedded link,

That width is determined by your browser's window size AFAIK. You could try [a CSS trick](https://stackoverflow.com/a/35153397) to keep aspect ratio of iframe instead of specifying arbitrary numbers.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml.js:
```diff
@@ -17,6 +17,8 @@ const rules = {
 	fountain: require('./MdToHtml/rules/fountain'),
 	mermaid: require('./MdToHtml/rules/mermaid').default,
+	sanitize_html: require('./MdToHtml/rules/sanitize_html').default,
+	video: require('./MdToHtml/rules/video'),
```
Please move this to the `plugins` object below because your plugin is setup like a regular markdown-it plugin, unlike the rules in this object.

Rishgod replied:
UPDATE: I moved `video: require('./MdToHtml/rules/video')`, back to the rules object in joplin-renderer/MdToHtml.js, (I went through other me rules and every one in which context was added, was under `rules `object so I changed its import which was done in `plugin `object to `rules `object)

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,29 @@
+
+	const defaultRender = markdownIt.renderer.rules.link_open;
+	const videoIDRegex = /^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|watch\?v=|&v=)([^#&?]*).*/;
+	const key = /.*(youtube|youtu.be).*/im;
```
Please could you change this to use only one regex. I think videoIDRegex should already give the ID so the `key` doesn't seem necessary.

Rishgod replied:
okay

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,29 @@
+
+		if (!match || !videoID) return defaultRender(tokens, idx, options, env, self);
+
+		if (match && videoID[2].length == 11) {
```
`videoID[2].length == 11` is not necessary if your regex is correct.

Rishgod replied:
it's just to prevent faulty youtube links, but I'll remove it if you say so

laurent22 replied:
> it's just to prevent faulty youtube links, but I'll remove it if you say so

It can't be faulty if the regex is correct. You could enforce a number of characters in your regex, although I don't think you should in case YouTube change the ID length.

Rishgod replied:
>It can't be faulty if the regex is correct. You could enforce a number of characters in your regex, although I don't think you should in case YouTube change the ID length.

okay I'll remove it

***

By faulty links, I meant if someone accidentally changes the video ID of the link(a typo)

![regex](https://user-images.githubusercontent.com/54576074/77446427-68826a00-6e14-11ea-88ab-c2dfaa6f8d00.gif)

It does seem a bit redundant, it was used in other youtube video-embeddings plugins that's why I added it. 


#### REGEX 

<img width="660" alt="regex" src="https://user-images.githubusercontent.com/54576074/77447158-5a811900-6e15-11ea-8bfa-67799c45274a.png">

PackElend replied:
where can I find this REGEX editor?

Rishgod replied:
https://www.debuggex.com/

Rishgod commented:
@bedwardly-down can you please review this, thank you

laurent22 commented:
@Rishgod, thanks for the update. I need to build and run the branch, but the code looks about right so we're nearly there.

laurent22 commented:
Thanks for the update @Rishgod. Now that I've ran the code, I've noticed two issues:

- Instead of rendering a link to a video, I think it should use the image markup for it. So  instead of `[](https://www.youtube.com/...)` it should be `![](https://www.youtube.com/...)`. Please could you change that? I believe you can simply use the `image` rule instead of the `link_open` one.

- Please make the video grow or shrink with the view, and set for example a max width of 560. Here's for example one method: https://www.h3xed.com/web-development/how-to-make-a-responsive-100-width-youtube-iframe-embed

- Please update the doc in readme/markdown.md, in particular the plugin table, which should have info on how to use your plugin.

laurent22 commented:
@Rishgod, now that I think about it, it would be best to use regular links (as you have now) to enable the plugin. This is because a regular link would still work in any Markdown viewer, while an image markup would break. Eventually we should use something like [Generic directives](https://talk.commonmark.org/t/generic-directives-plugins-syntax/444) but this is not part of the standard yet.

Rishgod commented:
Changes :
- made the iframe responsive. [Heres ](https://gph.is/g/apbpKqo)the demonstration
- Improved unit test by checking a variety of youtube links
- Moves file to import from plugin back to rules in joplin-renderer/MdToHtml.js, (I went through other me rules and every one in which CSS was added as a plugin so I reverted it back to the plugin)

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,61 @@
+// LOGIC: Extract the & UUID of the youtube video and render it in iframe
+const videoCss = () => {
+	return [
+		{
+			inline: true,
+			mime: 'text/css',
+			text: `
+                .video-container {
+                    position: relative;
+                    width: 100%;
+                    height: 0;
+                    padding-bottom: 56.25%;
+                }
+                .video-iframe {
+                    position: absolute;
+                    top: 0;
+                    left: 0;
+                    width: 100%;
+                    height: 100%;
+				}`,
+		},
+	];
```
Please fix the indentation (tabs only)

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,61 @@
+			inline: true,
+			mime: 'text/css',
+			text: `
+                .video-container {
+                    position: relative;
+                    width: 100%;
+                    height: 0;
+                    padding-bottom: 56.25%;
+                }
+                .video-iframe {
```
Please prefix the class names with "jop" eg `.jop-video-container`. This is because notes can contain arbitrary HTML, especially those imported from web clipper, and so we add a prefix to avoid any conflict.

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,61 @@
+                    position: relative;
+                    width: 100%;
+                    height: 0;
+                    padding-bottom: 56.25%;
```
What do we need this padding for?

Rishgod replied:
It will result in a `<div>` with height equal to 56.25% of the width of its container (a 16:9 aspect ratio).

To know what padding-bottom value to use for each aspect ratio we can divide the height by the width.

16:9 aspect ratio.
9 / 16 = .5625
We need a percentage, so we multiply by 100
.5625 * 100 = 56.25

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,61 @@
+			if (!('video' in context.pluginAssets)) {
+				addContextAssets(context);
+			}
+			return `<div class="video-container"><iframe  class="video-iframe"  width="560" height="315" src="https://www.youtube.com/embed/${videoID[2]}"  frameborder="0" allowfullscreen></iframe></div>`;
```
- Please remove the width and height as that should be controlled by your CSS now.
- Please remove the double/triple spaces for example around `class="video-iframe"` or `frameborder="0"`

laurent22 commented on readme/markdown.md:
```diff
@@ -154,3 +154,4 @@ Note that the functionality added by these plugins is not part of the CommonMark
 | [Insert](https://github.com/markdown-it/markdown-it-ins) | `++inserted++` | Transforms into `<ins>inserted</ins>` (<ins>inserted</ins>) | no |
 | [Multitable](https://github.com/RedBug312/markdown-it-multimd-table) | See [MultiMarkdown](https://fletcher.github.io/MultiMarkdown-6/syntax/tables.html) page | Adds more power and customization to markdown tables | no |
 | [Fountain](https://fountain.io) | <code>\`\`\`fountain</code><br/>Your screenplay...<br/><code>\`\`\`</code> | Adds support for the Fountain markup language, a plain text markup language for screenwriting | no |
+| [Video](https://github.com/laurent22/joplin/pull/2747) | A valid youtube URL. Eg:https://www.youtube.com/watch?v=uZClGt1xS1A| Adds support for embedding youtube video in Joplin's markdown editor. See [this](https://github.com/laurent22/joplin/pull/2747) for demonstration  | no |
```
For the description please use this: 

> Converts video URLs into a video player. Currently only YouTube is supported. More information in [this post](https://github.com/laurent22/joplin/pull/2747).

laurent22 commented on CliClient/tests/md_to_html/video.md:
```diff
@@ -0,0 +1,12 @@
+[](http://www.youtube.com/watch?v=iwGFalTRHDA)
+[](https://www.youtube.com/watch?v=iwGFalTRHDA) 
+[](https://www.youtube.com/watch?v=iwGFalTRHDA&feature=related)
+[](http://www.youtube.com/embed/watch?feature=player_embedded&v=iwGFalTRHDA)
+[](http://www.youtube.com/embed/watch?v=iwGFalTRHDA)
+[](http://www.youtube.com/watch?feature=player_embedded&v=iwGFalTRHDA)
+[](http://www.youtube.com/watch?v=iwGFalTRHDA)
+[](http://www.youtube.com/attribution_link?u=/watch?v=aGmiw_rrNxk&feature=sharea=9QlmP1yvjcllp0h3l0NwuA)
+[](http://www.youtube.com/attribution_link?a=fF1CWYwxCQ4&u=/watch?v=qYr8opTPSaQ&feature=em-uploademail)
+[](http://www.youtube.com/attribution_link?a=fF1CWYwxCQ4&feature=em-uploademail&u=/watch?v=qYr8opTPSaQ)
+https://www.youtube.com/watch?v=SADub7W22Zg
+[google](https://www.google.com/)
```
Thanks for the detailed tests. It also need to work with these cases:

- `[some title here](http://www.youtube.com/watch?v=iwGFalTRHDA)` => should create a video player and display a link "some title here" below the video, which opens the video on YouTube.
- `http://www.youtube.com/watch?v=iwGFalTRHDA` => should create a video player
- `https://www.a.com/watch?v=SADub7W22Zg` => should **not** create a video player (it currently does!)

Rishgod replied:
![image](https://user-images.githubusercontent.com/54576074/79128545-51a2b800-7dc1-11ea-89e5-a0f32193b289.png)

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,61 @@
+
+function installRule(markdownIt, mdOptions, ruleOptions, context) {
+	const defaultRender = markdownIt.renderer.rules.link_open;
+	const videoIDRegex = /^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|watch\?v=|&v=)([^#&?]*).*/;
```
This regex is faulty, for examples it matches this, but shouldn't:

`'https://www.a.com/watch?v=SADub7W22Zg'.match(/^.*(youtu\.be\/|v\/|u\/\w\/|embed\/|watch\?v=|&v=)([^#&?]*).*/)`

Rishgod replied:
Sure, we can remove this bug by checking if the domain name is youtube/youtu.be
``` js
	const key = /.*(youtube|youtu.be).*/im;
	const match = link.match(key);
if (videoID[2] && match[1]==='youtube') {...}
```

laurent22 replied:
How about looking at a url parsing lib? Or maybe there's already a function for this in Joplin?

Rishgod replied:
Okay, I'll look into it and let you know!

laurent22 replied:
You can use the `url` package, which is already available to all clients: https://www.npmjs.com/package/url

Rishgod replied:
yea I looked into it, and we currently use this dependency in Joplin too, this update got delayed a bit because I was busy trying to implement [this ](https://discourse.joplinapp.org/t/add-tables-to-the-toolbar/6213/15?u=rishabh.malhotra) feature request. I will get this done ASAP

laurent22 commented on ReactNativeClient/lib/models/Setting.js:
```diff
@@ -387,6 +387,7 @@ class Setting extends BaseModel {
 			'markdown.plugin.emoji': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['mobile', 'desktop'], label: () => `${_('Enable markdown emoji')}${wysiwygNo}` },
 			'markdown.plugin.insert': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['mobile', 'desktop'], label: () => `${_('Enable ++insert++ syntax')}${wysiwygNo}` },
 			'markdown.plugin.multitable': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['mobile', 'desktop'], label: () => `${_('Enable multimarkdown table extension')}${wysiwygNo}` },
+			'markdown.plugin.video': { value: false, type: Setting.TYPE_BOOL, section: 'plugins', public: true, appTypes: ['desktop'], label: () => `${_('Enable video embedding')}${wysiwygNo}` },
```
This works on mobile too as far as I can see so please enable it in this case too.

Rishgod replied:
okay

Rishgod commented:
- fixed the indentation (tabs only)
- prefixed the class names with "jop" eg `.jop-video-container` 
- removed typos 
- height and width in the iframe is removed(managed by CSS)
- updated the description in readme/markdown
- improved test cases & included the functionality for the additional test cases, that was asked for

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,70 @@
+		const token = tokens[idx];
+		const link = token.attrGet('href');
+		const videoID = link.match(videoIDRegex);
+		const hostname = url.parse(link).hostname;
```
If you use a url parser you no longer need your regex. Just print the result of the url.parse call and you'll see you have all the elements that you need.

Rishgod replied:
Sir since youtube supports a number of link format its preferable to just use regex, because if we use the passed URL method, we will end up checking for a lot of cases, let me explain
 
Youtube currently supports 4 types of link
|       |      |
| --- | --- |
|normal  http://www.youtube.com/watch?v=USjZcfj8yxE
shortened https://youtu.be/USjZcfj8yxE
no-cookie - http://www.youtube-nocookie.com/v/USjZcfj8yxE?version=3&hl=en_US&rel=0
legacy -

laurent22 replied:
Sorry but I don't want to go on and on about this URL parsing thing, this is a solved problem and there's already a parser in Joplin, so please use this.

I know you can use a regex, but yours had various issues and I personally can't make sense of it (as they said "he had a problem and tried to solve it with a regex, and now he has two problems). The parser does that job for you and has been battle tested in thousands of projects, so that's what we should use if we want something reliable.

Rishgod replied:
Alright, sir, I would implement this with URL parsing, but I seriously urge to read this,

the current regex had the bug you mentioned 
> https://www.a.com/watch?v=SADub7W22Zg => should not create a video player (it currently does!)

because the regex should only be used when you know that you are dealing with a youtube URL. Else it  caused false positives, e.g. this passes: 'v/2059-9080-5437'

So now that we do check that we are dealing with a youtube link it shouldn't have any problem.
you can check here
https://stackoverflow.com/a/8260383/12809166

Rishgod replied:
> Sorry but I don't want to go on and on about this URL parsing thing, this is a solved problem and there's already a parser in Joplin, so please use this.
> 
> I know you can use a regex, but yours had various issues and I personally can't make sense of it (as they said "he had a problem and tried to solve it with a regex, and now he has two problems"). The parser does that job for you and has been battle tested in thousands of projects, so that's what we should use if we want something reliable.

Sir I have made a change using the npm package [get-video-id](https://www.npmjs.com/package/get-video-id) used for extracting video ID for youtube and Vimeo, judging for the comment you made, you not keen on using this dependency and want to use the preexisting URL parser, but I seriously urge you to take a look at the changes.

Rishgod replied:
@laurent22  sir the extraction of youtube video ID is quite complex because you have to take into account for a variety of links, I tried to make a function which extracts the videoID but, it works fine for URLs which have video id in the query but URLs which don't have video id in querystring, then have to be extracted from path, and there you'll end up using some sort of regex to extract the video, I seriously believe the StackOverflow regex for getting videoID will work for 99% of the cases.
https://repl.it/@Rishgod/VideoID-extraction
_(I tried to implement it using parsing url)_

laurent22 commented on ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/video.js:
```diff
@@ -0,0 +1,70 @@
+		const link = token.attrGet('href');
+		const videoID = link.match(videoIDRegex);
+		const hostname = url.parse(link).hostname;
+		const checkHostName = validHostName.includes(hostname);
```
checkHostName => isValidHostname

laurent22 commented:
Closing pull request (PM sent to Rishgod)
